crates/debug/src/transform/expression.rs

```diff
@@ -465,23 +576,31 @@ where
         }
     }
 
-    if !code_chunk.is_empty() {
-        parts.push(CompiledExpressionPart::Code(code_chunk));
-    }
+    unread_bytes = 0;
+    flush_code_chunk!();
 
-    if base_len > 0 && base_len + 1 < parts.len() {
-        // see if we can glue two code chunks
-        if let [CompiledExpressionPart::Code(cc1), CompiledExpressionPart::Code(cc2)] =
-            &parts[base_len..=base_len]
-        {
-            let mut combined = cc1.clone();
-            combined.extend_from_slice(cc2);
-            parts[base_len] = CompiledExpressionPart::Code(combined);
-            parts.remove(base_len + 1);
-        }
-    }
+    parts.push(CompiledExpressionPart::LandingPad {
+        original_pos: expr.0.len().into_u64() as usize,
+    });
 
-    Ok(Some(CompiledExpression { parts, need_deref }))
+    parts.push(CompiledExpressionPart::Code(vec![])); // so that we have enough windows below
+
+    // collect all CompiledExpressionParts but only relevant LandingPads:
+    //  - those followed by a Code chunk that is being jumped into
+    //  - those which are either at the start or end of a jump arc
+    let parts = parts
+        .windows(2)
+        .filter_map(|p| match p {
+	    [CompiledExpressionPart::LandingPad { original_pos }, CompiledExpressionPart::Code(code_chunk)]
+		if jump_arcs.values().any(|t| (original_pos + 1..original_pos + code_chunk.len()).contains(t))
+		=> Some(p[0].clone()),
+	    [CompiledExpressionPart::LandingPad { original_pos }, _]
+		if jump_arcs.iter().all(|(from, to)| from != original_pos && to != original_pos)
+		=> None,
+	    _ => Some(p[0].clone())
+	})
+	.collect();
+    Ok(Some((CompiledExpression { parts, need_deref }, jump_arcs)))
```

Comment:
Let's make `jump_arcs` as field of `CompiledExpression` (`from_label` will have empty map)